txn/txn.go

```diff
@@ -397,20 +398,33 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 		iter := c.Find(nil).Select(bson.M{"_id": 1, "txn-queue": 1}).Iter()
 		var tdoc TDoc
 		for iter.Next(&tdoc) {
+			idsToRemove := make(map[string]bool)
 			for _, fullTxnId := range tdoc.TxnIds {
 				txnId := bson.ObjectIdHex(fullTxnId[:24])
 				if found[txnId] {
 					continue
 				}
-				if r.tc.FindId(txnId).One(nil) == nil {
-					found[txnId] = true
-					continue
-				}
-				logf("WARNING: purging from document %s/%v the missing transaction id %s", collection, tdoc.DocId, txnId)
-				err := c.UpdateId(tdoc.DocId, M{"$pull": M{"txn-queue": M{"$regex": "^" + txnId.Hex() + "_*"}}})
-				if err != nil {
-					return fmt.Errorf("error purging missing transaction %s: %v", txnId.Hex(), err)
+				if !knownMissing[txnId] {
```

Comment:
I'm not convinced that caching knownMissing is going to help much. For any given transaction, the number of miss counts will only be as many as the number of documents affected by that transaction, which is going to be pretty small (and in our case is only going to be one). That said, we've already got a map appropriate for storing both found and not found - how about using found for both, storing false when we've looked up an id and found it to be missing?

I think we'd get a much better speedup by making the id finding concurrent.